xmppserver/src/main/java/org/jivesoftware/openfire/muc/spi/LocalMUCRoom.java

```diff
@@ -792,41 +792,47 @@ private void sendInitialPresences(LocalMUCRole joinRole) {
     }
 
     public void occupantAdded(OccupantAddedEvent event) {
-        // Create a proxy for the occupant that joined the room from another cluster node
-        RemoteMUCRole joinRole = new RemoteMUCRole(mucService, event);
-        JID bareJID = event.getUserAddress().asBareJID();
-        String nickname = event.getNickname();
-        List<MUCRole> occupants = occupantsByNickname.get(nickname.toLowerCase());
-        // Do not add new occupant with one with same nickname already exists
-        if (occupants == null) {
-            occupants = new ArrayList<>();
-            occupantsByNickname.put(nickname.toLowerCase(), occupants);
-        } else {
-            // sanity check; make sure the nickname is owned by the same JID
-            if (occupants.size() > 0) {
-                JID existingJID = occupants.get(0).getUserAddress().asBareJID();
-                if (!bareJID.equals(existingJID)) {
-                    Log.warn(MessageFormat.format("Conflict detected; {0} requested nickname '{1}'; already being used by {2}", bareJID, nickname, existingJID));
-                    return;
+        lock.writeLock().lock();
+        try {
+            // Create a proxy for the occupant that joined the room from another cluster node
+            RemoteMUCRole joinRole = new RemoteMUCRole(mucService, event);
+            JID bareJID = event.getUserAddress().asBareJID();
+            String nickname = event.getNickname();
+            List<MUCRole> occupants = occupantsByNickname.get(nickname.toLowerCase());
```

Comment:
Not as simple as line 819+, but this could be simplified with 

`List<MUCRole> existingOccupantsForNickname = occupantsByNickname.computeIfAbsent(nickname.toLowerCase, nickname->new ArrayList());
...`

Reply:
because of the return Statement at https://github.com/igniterealtime/Openfire/pull/1429/files/06bc632d69818756e603c229464c4c763af9e0e1#diff-e7b0858be35ba3ff73325d038d049ed5R812 it can not be done in a single "compute" without local variables or throwing a exception.